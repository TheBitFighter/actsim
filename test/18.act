deftype memiface (chan?(bool) rd; chan?(int) addr, din; chan!(int) dout) { }

template<pint N>
defproc mem(memiface i)
{
  int x[N];
  int a;
  bool b;
  chp {
    *[ log ("hi");
     [i.rd-> log("here"); i.addr?a, i.rd?; i.dout!x[a]
     []~i.rd -> log("there"); i.addr?a, i.rd?; i.din?x[a]
     ]]
  }
}

defproc test()
{
  mem<32> m;
  int x;

  chp {
     log("test mem");
     m.i.rd!false,m.i.addr!0,m.i.din!5;
     m.i.rd!false,m.i.addr!1,m.i.din!6;
     m.i.rd!true,m.i.addr!0,m.i.dout?x;
     log("got ", x);
     m.i.rd!true,m.i.addr!1,m.i.dout?x;
     log("got ", x)
   }
}
   
test t;
