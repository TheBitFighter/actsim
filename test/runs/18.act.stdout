[      10 t#:0] <> test mem
[      10 t#:0] <m> hi
[      20 t#:0] <m> there
[      50 t#:0] <m> hi
[      60 t#:0] <m> there
[      90 t#:0] <m> hi
[     100 t#:0] <m> here
[     130 t#:0] <> got 5
[     130 t#:0] <m> hi
[     140 t#:0] <m> here
[     170 t#:0] <> got 6
[     170 t#:0] <m> hi
